File: ExclusiveTimeFunctions.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct  8 10:23:09 2019

@author: tanvirkaur
Time Complexity = O(logs)
Space Complexity = O(n)
"""


class Solution(object):
    def exclusiveTime(self, n, logs):
        """
        :type n: int
        :type logs: List[str]
        :rtype: List[int]
        """
        res = [0]*n
        stack = []
        prev = 0
        for log in logs:
            first ,second,third  = log.split(':')
            first = int(first)
            third = int(third)
            curr = third
            if second == 'start':
                if stack:
                    res[stack[-1]] += curr - prev
                stack.append(first)
                prev = curr
            else:
                res[stack.pop()] += curr-prev + 1
                prev = curr + 1
        return res

File: ExclusiveTimeFunctions.cpp
// Time Complexity -> O(n)
// Space Complexity -> O(n) for the stack.
// Problems Faced - No!
// It runs on Leetcode!
class Solution {
    private:
    vector<string> splitLog(string s){
        vector<string> result;
        int l = s.size();
        string curr = "";
        for(int i = 0; i < l; i++){
            if(s[i] != ':')
                curr += s[i];
            if(s[i] == ':' || i == l-1){
                if(curr != ":")
                    result.push_back(curr);
                curr = "";
            }
        }
        return result;
    }
public:
    vector<int> exclusiveTime(int n, vector<string>& logs) {
        vector<int> answer(n);
        stack<int> st;
        int prev = 0;
        for(string log : logs){
            vector<string> currLog = splitLog(log);
            int id = stoi(currLog[0]);
            int timeSt = stoi(currLog[2]);
            if(currLog[1] == "start"){
                if(!st.empty()){
                    int prevId = st.top();
                    answer[prevId] +=  timeSt - prev;
                    
                }
                st.push(id);
                prev = timeSt;
            }else{
                int prevId = st.top(); st.pop();
                answer[prevId] += timeSt + 1 - prev;
                prev = timeSt+1;
            }
        }
        return answer;
    }
};